api/src/middleware/connected-user-log.js

```diff
@@ -0,0 +1,15 @@
+const auth = require('../auth');
+const connectedUserLogService = require('../services/connected-user-log');
+const logger = require('../logger');
+
+module.exports = {
+  log: (req, res, next) => {
+    return auth
+      .getUserCtx(req)
+      .then(({ name }) => {
+        connectedUserLogService.save(name);
+      })
+      .catch(err => logger.error('Error recording user connection: %o', err))
+      .then(next);
```

Comment:
This is still causing problems... I checked the code out locally and modified the connected user log service to always reject. When I logged in I saw...

```
2021-06-09 11:29:10 ERROR: UNHANDLED REJECTION! 
2021-06-09 11:29:10 ERROR:   Reason: Error: boom
    at Object.saveLog [as save] (/home/gareth/src/medic/cht-core/api/src/services/connected-user-log.js:7:25)
    at /home/gareth/src/medic/cht-core/api/src/middleware/connected-user-log.js:10:33
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  [stack]: 'Error: boom\n' +
    '    at Object.saveLog [as save] (/home/gareth/src/medic/cht-core/api/src/services/connected-user-log.js:7:25)\n' +
    '    at /home/gareth/src/medic/cht-core/api/src/middleware/connected-user-log.js:10:33\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:93:5)',
  [message]: 'boom'
} 
```

Unhandled rejections like that are always a mistake (hence the app caps!), so we definitely need to return the promise from of the connected user service call.

When I added the return here I see the server error you're talking about, which is also not cool! Because of the shorthand used, the `catch` on line 12 is actually returning the result of the logger call, which is automatically being provided as the first parameter of the `next` function which is causing the error.

However, this code worked perfectly...

```suggestion
    return auth
      .getUserCtx(req)
      .then(({ name }) => connectedUserLogService.save(name)) // return the promise so we don't get unhandled rejections
      .catch(err => { // don't use the single line version because it auto-returns and we don't want to return the logger object
        logger.error('Error recording user connection: %o', err);
      })
      .then(() => next()); // explicitly call `next()` so we can guarantee there are no parameters being passed in automatically
```

Feel free to remove or improve the inline comments - I added those to explain my thinking to you.

Reply:
Thanks for explaining your thinking. Very insightful!